ssas/service/public/api_test.go

```diff
@@ -462,3 +466,444 @@ func (s *APITestSuite) TestJsonError() {
+	//System does not need to be created for this test since header and param checks are done before assertion is parsed/looked up.
+	form := url.Values{}
+	form.Set("scope", "system/*.*")
+	form.Set("client_assertion_type", "urn:ietf:params:oauth:client-assertion-type:jwt-bearer")
+	form.Set("client_assertion", "value_does_not_matter_for_this_test")
+	handler := http.HandlerFunc(tokenV2)
+
+	//Invalid grant_type param
+	form.Set("grant_type", "invalid_grant_type")
+	req := httptest.NewRequest("POST", "/v2/token", strings.NewReader(form.Encode()))
+	req.Header.Add("Accept", "application/json")
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+	handler.ServeHTTP(s.rr, req)
+	s.verifyErrorResponse(http.StatusBadRequest, "invalid value for grant_type")
+}
+
+func (s *APITestSuite) TestClientAssertionAuthWithBadClientAssertionTypeParam() {
+	//System does not need to be created for this test since header and param checks are done before assertion is parsed/looked up.
+	form := url.Values{}
+	form.Set("scope", "system/*.*")
+	form.Set("grant_type", "client_credentials")
+	form.Set("client_assertion", "value_does_not_matter_for_this_test")
+	handler := http.HandlerFunc(tokenV2)
+
+	//Invalid client_assertion_type param
+	form.Set("client_assertion_type", "invalid_client_assertion_type")
+	req := httptest.NewRequest("POST", "/v2/token", strings.NewReader(form.Encode()))
+	req.Header.Add("Accept", "application/json")
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+	handler.ServeHTTP(s.rr, req)
+	s.verifyErrorResponse(http.StatusBadRequest, "invalid value for client_assertion_type")
+}
+
+func (s *APITestSuite) TestClientAssertionAuthWithMissingClientAssertionParam() {
+	//Create system and valid client assertion token
+	form := url.Values{}
+	form.Set("scope", "system/*.*")
+	form.Set("grant_type", "client_credentials")
+	form.Set("client_assertion_type", "urn:ietf:params:oauth:client-assertion-type:jwt-bearer")
+	handler := http.HandlerFunc(tokenV2)
+
+	//Missing client_assertion param
+	req := httptest.NewRequest("POST", "/v2/token", strings.NewReader(form.Encode()))
+	req.Header.Set("Accept", "application/json")
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+
+	handler.ServeHTTP(s.rr, req)
+	s.verifyErrorResponse(http.StatusBadRequest, "missing client_assertion")
+}
+
+func (s *APITestSuite) verifyErrorResponse(expectedStatus interface{}, expectedMsg string) {
+	assert.Equal(s.T(), expectedStatus, s.rr.Code)
+	t := ssas.ErrorResponse{}
+	assert.NoError(s.T(), json.NewDecoder(s.rr.Body).Decode(&t))
+	assert.NotEmpty(s.T(), t)
+	assert.Equal(s.T(), expectedMsg, t.Error)
+}
+
+func mintClientAssertion(issuer string, subject string, aud string, issuedAt int64, expiresAt int64, privateKey *rsa.PrivateKey) (*jwt.Token, string, error) {
+	claims := service.CommonClaims{}
+
+	token := jwt.New(jwt.SigningMethodRS512)
+	tokenID := uuid.NewRandom().String()
+	claims.TokenType = "ClientAssertion"
```

Comment:
I don't think we're going to have users set the TokenType claim.  Shall we remove this line?